desc/integrals/_bounce_utils.py

```diff
@@ -700,127 +845,315 @@ def interp_to_argmin_hard(h, points, knots, g, dg_dz, method="cubic"):
+        First (second) element stores left (right) boundaries.
+    knots : jnp.ndarray
+        Shape (knots.size, ).
+        z coordinates of spline knots. Must be strictly increasing.
+    g : jnp.ndarray
+        Shape (..., knots.size - 1, g.shape[-1]).
+        Polynomial coefficients of the spline of g in local power basis.
+        Last axis enumerates the coefficients of power series. Second to
+        last axis enumerates the polynomials that compose a particular spline.
+    dg_dz : jnp.ndarray
+        Shape (..., knots.size - 1, g.shape[-1] - 1).
+        Polynomial coefficients of the spline of ∂g/∂z in local power basis.
+        Last axis enumerates the coefficients of power series. Second to
+        last axis enumerates the polynomials that compose a particular spline.
+    beta : float
+        More negative gives exponentially better approximation at the
+        expense of noisier gradients - noisier in the physics sense (unrelated
+        to the automatic differentiation).
+    upper_sentinel : float
+        Something larger than g. Choose value such that
+        exp(max(g)) << exp(``upper_sentinel``). Don't make too large or numerical
+        resolution is lost.
+
+    Warnings
+    --------
+    Recall that if g is small then the effect of β is reduced.
+    If the intention is to use this function as argmax, be sure to supply
+    a lower sentinel for ``upper_sentinel``.
 
     Returns
     -------
-    fig, ax
-        Matplotlib (fig, ax) tuple.
+    h : jnp.ndarray
+        Shape (..., num well).
 
     """
-    fig, ax = plt.subplots()
-    legend = {}
-    if include_knots:
-        for knot in ppoly.x:
-            _add2legend(
-                legend,
-                ax.axvline(
-                    x=knot, color="tab:blue", alpha=knot_transparency, label="knot"
-                ),
-            )
+    ext, g_ext = _get_extrema(knots, g, dg_dz, sentinel=0)
+    # Our softargmax(x) does the proper shift to compute softargmax(x - max(x)),
+    # but it's still not a good idea to compute over a large length scale, so we
+    # warn in docstring to choose upper sentinel properly.
+    argmin = softargmax(
+        beta * _where_for_fft_argmin(points, ext, g_ext, upper_sentinel),
+        axis=-1,
+    )
+    return jnp.linalg.vecdot(
+        argmin,  # shape is (..., num well, num extrema)
+        interp_rfft2(
+            T.eval1d(ext),
+            ext,
+            h[..., jnp.newaxis, :, :],
+            domain1=(0, 2 * jnp.pi / NFP),
+            axes=(-1, -2),
+        )[..., jnp.newaxis, :],
+        # adding axis to broadcast with num well axis
+    )
 
-    z = jnp.linspace(
-        start=setdefault(start, ppoly.x[0]),
-        stop=setdefault(stop, ppoly.x[-1]),
-        num=num,
+
+# TODO: Generalize this beyond ζ = ϕ or just map to Clebsch with ϕ.
+def get_fieldline(alpha_0, iota, num_transit, period):
+    """Get sequence of poloidal coordinates A = (α₀, α₁, …, αₘ₋₁) of field line.
```

Comment:
this is specifically, if one were to follow a field line 2pi/NFP around (or whatever is given by period), and then re-label the field line with the alpha there assuming phi=0, these are the alpha values?

slightly more explanaiotn could be useful here I think